### PR TITLE
Docker Development Environment

### PR DESCRIPTION
**Pull request by fabianhauser**
It would be nice to have a Docker development environment as per #189 and, if possible, also a production Docker deployment to make things easier and consistent.

This pull request is mainly for tracking development progress and discussing implementation issues, so please wait with merging until we have a stable version.
 
Please let me now if you have ideas / improvements! :wink:

### CODE REVIEW

**Comment by fabianhauser** [comment]
Currently in progress:

- Basic Docker & Docker Compose structure
- Collecting all dependencies from the existing Ansible scripts

File: Dockerfile-base [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+
+#TODO: Do we really need build-essential here?
+RUN apt-get update && \
+  apt-get -y install \
+    build-essential gcc g++ \
+    python-dev python-setuptools python-pip python-virtualenv \
+    git mercurial subversion \
+    postgres-client libpq-dev \
+    libjpg-progs libjpg-dev imagemagick \
+    ruby ruby-dev \
```
Comment (dbrgn) [inline Dockerfile-base]: The Ruby stuff could probably be replaced by a pyscss pipeline in the fuutre.

File: Dockerfile-base [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+RUN gem install compass
+
+RUN groupadd -g 1000 studentenportal
+RUN useradd --home /home/studentenportal -u 1000 -g 1000 -M studentenportal
+
+RUN mkdir -p /srv/www/studentenportal && chown -R studentenportal:studentenportal /srv/www/
+
+WORKDIR /srv/www/studentenportal
+USER studentenportal
+CMD [ "python", "manage.py", "runserver" ]
```
Comment (dbrgn) [inline Dockerfile-base]: I'm not sure this should be in the base Dockerfile. This way to start the server should only be used in development.

File: docker-compose-production.yml [hunk 3]
```diff
@@ -0,0 +1,10 @@
+#  
+# studentenportal ----- studentenportal-media:/var/www/studentenportal/media
+#  ||__
+#  | __ postgres ----- postgres-data:/var/lib/postgresql/data
+#  ||__
+#  | __ mailserver #TODO: postfix?
```
Comment (dbrgn) [inline docker-compose-production.yml]: Honestly I'd use something like https://www.sparkpost.com/ if possible.

File: Dockerfile-base (hunk 1 above)
Comment (saspeed) [inline Dockerfile-base]: Running the docker build of the current Dockerfile fails installing standard packages
```
...
Get:7 http://deb.debian.org jessie/main amd64 Packages [9064 kB]
Fetched 9848 kB in 31s (311 kB/s)
Reading package lists...
Reading package lists...
Building dependency tree...
Reading state information...
E: Unable to locate package postgres-client
E: Unable to locate package libjpg-progs
E: Unable to locate package libjpg-dev
```
Any suggestions on wha I am missing here?

**Comment by raphiz** [comment]
I was thinking about security patches.
If we use prebuild images such as postgress we must patch every container manually - which is a tedious task, especially since different versions of different distributions _might_ be in use (and we can't control it fully). The downside is that we can't use existing containers like postgres.

I therefore would suggest to build one base image, extending from a minimal Debian or Ubuntu base image. This would be the place to apply any critical patches. If we have a proper deployment process, we can simply rebuild all images, rebuild the containers and have the patches applied everywhere.

**Comment by dbrgn** [comment]
> If we use prebuild images such as postgress we must patch every container manually

Not really. If you use `docker-compose up -d` regularly, I think all containers should be recreated if the base image was updated.

Alternatively we can also deploy them directly with ansible. The ansible docker module has an option to always re-fetch and rebuild containers on each deploy.

**Comment by fabianhauser** [comment]
- [ ] Layout-Breite ist kaputt
- [ ] Layout für Zusammenfassung ist kaputt
- [ ] Vorschaubild für Zusammenfassung ist kaputt:  image

**Comment by fabianhauser** [comment]
TODO: 
- [ ] Testdaten
- [ ] Testing
  - [ ] Tests fixen in Docker Container
  - [ ] Ausführen von Tests vereinfachen.

File: Makefile [hunk 4]
```diff
@@ -33,5 +33,5 @@ clean-dev: stop-dev
 clean-production: stop-production
 	docker-compose --file docker-compose-production.yml rm
 
-tests: start-dev
-	docker-compose exec studentenportal-dev ./test.sh
+tests:
+	docker-compose run studentenportal-dev ./test.sh
```
Comment (raphiz) [inline Makefile]: Wenn so, dann aber mit `--rm`

**Comment by fabianhauser** [comment]
Hey @saspeed und @dbrgn 
Wir sind soweit fertig mit der docker-dev-env. Dabei hat es u.a. beim Design & README ein paar Anpassungen gegeben. Könntet ihr das ganze mal testen?

**Pull request merged**